File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<int> nodesBetweenCriticalPoints(ListNode* head) {
        vector<int> arr;
        ListNode* prev = head, *curr = head->next;
        int index = 1;
        while(curr && curr->next) {
            if(curr->val > prev->val && curr->val > curr->next->val)
                // local maxima
                arr.push_back(index);
            else if(curr->val < prev->val && curr->val < curr->next->val)
                // local minima
                arr.push_back(index);
            
            prev = curr;
            curr = curr->next;
            index++;
        }
        
        int n = arr.size();
        if(n < 2)
            return {-1, -1};
        int maxDist = arr[n-1] - arr[0];
        int minDist = INT_MAX;
        for(int i=1; i<n; i++)
            minDist = min(minDist, arr[i] - arr[i-1]);
        
        return {minDist, maxDist};
    }
};

File: 2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points/2058-find-the-minimum-and-maximum-number-of-nodes-between-critical-points.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int[] nodesBetweenCriticalPoints(ListNode head) {
        int[] ans = new int[] { Integer.MAX_VALUE, Integer.MIN_VALUE }; // [minDist, maxDist]

        ListNode prev = head;
        ListNode curr = head.next;
        ListNode next = curr.next;

        if (next == null) {
            return new int[] { -1, -1 };
        }

        boolean found = false;
        int distFromFirst = 0;
        int distFromPrev = 0;

        while (next != null) {
            if (found) {
                distFromFirst++;
                distFromPrev++;
            }
            if (isCritical(prev.val, curr.val, next.val)) {
                if (found) {
                    ans[0] = Math.min(ans[0], distFromPrev);
                    ans[1] = Math.max(ans[1], distFromFirst);
                    distFromPrev = 0;
                } else {
                    found = true;
                }
            }

            prev = curr;
            curr = next;
            next = next.next;
        }

        if (ans[0] == Integer.MAX_VALUE || ans[1] == Integer.MIN_VALUE) {
            return new int[] { -1, -1 };
        }

        return ans;
    }

    private boolean isCritical(int prev, int curr, int next) {
        if ((curr < prev && curr < next) ||
                (curr > prev && curr > next)) {
            return true;
        }

        return false;
    }
}